src/caf/core/data_structures.py

```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
+                    )
+        return targets
+
+    def ipf(self, targets: list[IpfTarget], tol: float = 1e-5, max_iters: int = 100):
+        """
+        Implement iterative proportional fitting for DVectors.
+
+        Parameters
+        ----------
+        targets: list[IpfTarget]
+            A list of targets to try and match. Due to the nature of the process,
+            the final target in this list will be matched most closely to. This
+            doesn't matter much when a good convergence is achieved, but can have
+            a large impact when not.
+        tol: float = 1e-5
+            The RMSE between self and targets the process will attempt to reach
+            before stopping. Can also stop if consecutive iterations score the same.
+        max_iters: int = 100
+            The max number of iterations which can run before the process will stop.
+
+        Returns
+        -------
+        DVector matched to targets, rmse achieved.
+        """
+        # check DVectors compatible
+        targets = self.validate_ipf_targets(targets)
+        new_dvec = self.copy()
+        prev_rmse = np.inf
+        bypass = False
+        for i in range(max_iters):
+            for target in targets:
+                if i > 0:
+                    bypass = True
+                inner = new_dvec.copy()
+                if target.segment_translations is not None:
+                    for targ_seg, seed_seg in target.segment_translations.items():
+                        inner = inner.translate_segment(
+                            from_seg=seed_seg, to_seg=targ_seg, _bypass_validation=bypass
+                        )
+                agg = inner.aggregate(target.data.segmentation, _bypass_validation=bypass)
+                if target.zone_translation is not None:
+                    agg = agg.translate_zoning(
+                        target.data.zoning_system,
+                        trans_vector=target.zone_translation,
+                        _bypass_validation=bypass,
+                    )
+                factor = target.data.__truediv__(agg, _bypass_validation=bypass)
+                if target.zoning_diff:
+                    factor = factor.translate_zoning(
+                        self.zoning_system,
+                        trans_vector=target.zone_translation,
+                        one_to_one=True,
+                        check_totals=False,
+                    )
+                if target.segment_translations is not None:
+                    for targ_seg, seed_seg in target.segment_translations.items():
+                        factor = factor.translate_segment(
+                            from_seg=targ_seg,
+                            to_seg=seed_seg,
+                            reverse=True,
+                            _bypass_validation=bypass,
+                        )
+                new_dvec = new_dvec.__mul__(factor, _bypass_validation=bypass)
```

Comment:
Method for factoring to targets all looks good, although there is lots of translation happening which keeps creating new DVectors and copying data is this slowing things down at all? In future is it worth looking into any ways to potentially reduce the amount of translations required or potentially performing some calculations outside the loop which can help with them, if speed isn't a problem then not worth bothering but possibly an area to revisit in future if speed is an issue

Reply:
I do think this may be a problem but at the moment I think all translations are necessary. Possibly the final translations of segment/zoning could be removed (getting factors back to seed target/zoning) if the seed segment(s) and zoning were kept in the DVector when throughout. For segments this would be relatively easy, but for zoning this would mean a refactor to allow multiindexed columns and multiple zone systems in a DVector. I have refactored so that after the first iteration there is no validation on DVector creation which should speed it up a bit.